rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+# RESTful Service Specifications
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Goals
+
+ - Uniformity between data-centric APIs and remote procedure APIs
+ - Use of pre-existing standards/conventions when possible
+ - Minimization of bike-shedding
+
+Standards/Conventions used as inspiration:
+ - [Whitehouse API Standards](https://github.com/WhiteHouse/api-standards)
+  - RESTful URLs
+ - [Spring Data REST](http://docs.spring.io/spring-data/rest/docs/current/reference/html/)
+  - paging conventions
+  - uses [JSON HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08)
+ - [JSend](http://labs.omniti.com/labs/jsend)
+  - API response conventions
+
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
+
+The base URI for a given resource type is based on the *singular* name of the
+resource. Example:
+
+    https://example.com/rest/v1/user
+
+## Operations Against a Resource Type
+
+The following operations may be supported:
+
+| HTTP Operation | Description            | Query Parameters              | Request Body | Response Body  |
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
+| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
+
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
```

Comment:
I think what you guys are saying is great. In fact, that'd be my preference @BVulaj @kahowell. But I would like to make sure that the design addresses or at least acknowledges as a weakness all issues on the table.  

Currently it doesn't address the following two issues.  
**Encoding issues**
> We have had problems with path params in the past, specifically that path
params are not encoded automatically by libraries as they are for query
params.
I'd like to hear more detail on @jewzaam.  I feel like having consistent numerical identifiers or UUIDs would be fine for this.

**Translation issues**
> I think this line of reasoning can be risky long term. It is good to strive for ubiquitous language between developers, QA, and business users. If we speak the same language, there is less lost in translation. The identifiers for a domain "thing" can be important, so I think it's not unreasonable to ask everyone to speak the same language here, though maybe not critical for ids. I would empathize with QA here particularly, though.

Although I still believe that business users shouldn't be concerned about API implementation, I do understand that such translation failures can lead to requirements replete with metric mixup mines.  I think the business users, if they truly are using/writing requirements for the API, will get used to having a different identifier used in APIs.